packages/filesystem/src/browser/file-resource.ts

```diff
@@ -35,30 +39,55 @@ export class FileResource implements Resource {
     }
 
     async readContents(options?: { encoding?: string }): Promise<string> {
+        this.state = await this.doReadContents(options);
+        return this.state.content;
+    }
+    protected async doReadContents(options?: { encoding?: string }): Promise<FileResource.State> {
         try {
-            const { stat, content } = await this.fileSystem.resolveContent(this.uri.toString(), options);
-            this.stat = stat;
-            return content;
+            if (!await this.fileSystem.exists(this.uriString)) {
+                return FileResource.emptyState;
+            }
+            if (await this.isOutdated()) {
+                return this.fileSystem.resolveContent(this.uriString, options);
+            }
+            return this.state;
         } catch {
-            return '';
+            return FileResource.emptyState;
         }
     }
 
-    async saveContents(content: string, options?: { encoding?: string }): Promise<void> {
-        const uri = this.uri.toString();
-        if (await this.fileSystem.exists(uri)) {
-            this.stat = this.stat || await this.fileSystem.getFileStat(this.uri.toString());
-            this.stat = await this.fileSystem.setContent(this.stat, content, options);
-        } else {
-            await this.fileSystem.createFile(uri, { content });
+    protected async isOutdated(): Promise<boolean> {
+        if (!this.state.stat) {
+            return true;
         }
+        const stat = await this.fileSystem.getFileStat(this.uriString);
+        /*
+         * if timestamps equal we cannot assume that content has not been changed because of
+         * https://nodejs.org/en/docs/guides/working-with-different-filesystems/#timestamp-resolution
+         */
+        return stat.lastModification >= this.state.stat.lastModification;
```

Comment:
See https://github.com/theia-ide/theia/pull/1188#issuecomment-362659715 if we are the only ones to save then the timestamp will always be equal... so this isOutdated is always true.

Should we check the contents with a sha1 for example instead or a md5 ?

Reply:
It is turned out that [my assumptions](https://github.com/theia-ide/theia/pull/1188#issuecomment-362595110) were wrong and this change does not fix an issue with the editor. I've decided to leave it since in some cases it can help. Adding hashes probably can improve even more, but again when hashes equal we cannot assume equality of content and it will be a case if one call read immediately after save.

I will look for the root cause of the original issue next week, but I don't think this code related.

Reply:
If you use hash you can know that equally is the same content that's the point of using a hash no ?

Reply:
Equality of hashes does not guarantee equality of content. The point is that when hashes are not equal you know for sure that content also, and can spare yourself from the content comparison.

Reply:
@akosyakov that's not true. Doing a SHA-256 on a file will guarantee equality of content.  Collisions can happen in theory but are very very unlikely in our use case.

Reply:
ok, we can have a look at it separately

Reply:
I've removed outdated check